### PR TITLE
Debug symbols: GDB session

### PR DESCRIPTION
**Pull request by Karm**
* Automates GDB sessions for local and in-container debugging.
 * Vertx/Mutiny Quarkus example with PostgresDB in a container
 * text parsing example to compare Native svm vs. Hotspot  
 * fixed windows support bugs (stopping containers, container logs)

### CODE REVIEW

**Comment by Karm** [comment]
# Testsuite (no builder image usage)

```
karm@localhost:~/workspaceRH/mandrel-integration-tests (debug-symbols *%)$  mvn clean verify -Ptestsuite  -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel:20.2.0.0.Final-java11
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] Native image integration TS                                        [pom]
[INFO] testsuite                                                          [jar]
[INFO] 
[INFO] ----------------< org.graalvm.tests.integration:parent >----------------
[INFO] Building Native image integration TS 1.0.0-SNAPSHOT                [1/2]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ parent ---
[INFO] 
[INFO] --------------< org.graalvm.tests.integration:testsuite >---------------
[INFO] Building testsuite 1.0.0-SNAPSHOT                                  [2/2]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ testsuite ---
[INFO] Deleting /home/karm/workspaceRH/mandrel-integration-tests/testsuite/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ testsuite ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ testsuite ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /home/karm/workspaceRH/mandrel-integration-tests/testsuite/target/classes
[INFO] 
[INFO] --- build-helper-maven-plugin:3.0.0:add-test-source (add-integration-test-source-as-test-sources) @ testsuite ---
[INFO] Test Source directory: /home/karm/workspaceRH/mandrel-integration-tests/testsuite/src/it/java added.
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ testsuite ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 0 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ testsuite ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 13 source files to /home/karm/workspaceRH/mandrel-integration-tests/testsuite/target/test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.22.2:test (default-test) @ testsuite ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.graalvm.tests.integration.RuntimesSmokeTest
2020-11-11 23:46:16.954 WARN  [o.g.t.i.u.Commands] (getProperty) Failed to detect any of QUARKUS_NATIVE_CONTAINER_RUNTIME,quarkus.native.container-runtime as env or sys props, defaulting to docker
2020-11-11 23:46:16.990 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Testing app: HELIDON_QUICKSTART_SE
2020-11-11 23:47:48.824 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Running...
2020-11-11 23:47:48.850 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Testing web page content...
2020-11-11 23:47:48.857 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Terminate and scan logs...
2020-11-11 23:47:49.028 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Gonna wait for ports closed...
2020-11-11 23:47:49.051 INFO  [o.g.t.i.u.Logs] (checkLog) build-and-run.log log for helidonQuickStart contains whitelisted error: `[WARNING] Warning: Ignoring server-mode native-image argument --no-server.'
2020-11-11 23:47:49.072 INFO  [o.g.t.i.u.Logs] (logMeasurements) 
App,buildTimeMs,timeToFirstOKRequestMs,RSSKb,FDs
HELIDON_QUICKSTART_SE,91818,20,62840,60
2020-11-11 23:47:49.098 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Testing app: QUARKUS_FULL_MICROPROFILE
2020-11-11 23:49:50.334 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Running...
2020-11-11 23:49:50.361 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Testing web page content...
2020-11-11 23:49:51.238 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Terminate and scan logs...
2020-11-11 23:49:51.390 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Gonna wait for ports closed...
2020-11-11 23:49:51.396 INFO  [o.g.t.i.u.Logs] (logMeasurements) 
App,buildTimeMs,timeToFirstOKRequestMs,RSSKb,FDs
QUARKUS_FULL_MICROPROFILE,121236,26,89736,107
2020-11-11 23:49:51.425 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Testing app: MICRONAUT_HELLOWORLD
2020-11-11 23:51:20.525 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Running...
2020-11-11 23:51:20.656 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Testing web page content...
2020-11-11 23:51:20.656 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Terminate and scan logs...
2020-11-11 23:51:20.808 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Gonna wait for ports closed...
2020-11-11 23:51:20.810 INFO  [o.g.t.i.u.Logs] (checkLog) build-and-run.log log for micronautHelloWorld contains whitelisted error: `[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.'
2020-11-11 23:51:20.810 INFO  [o.g.t.i.u.Logs] (checkLog) build-and-run.log log for micronautHelloWorld contains whitelisted error: `[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.'
2020-11-11 23:51:20.811 INFO  [o.g.t.i.u.Logs] (checkLog) build-and-run.log log for micronautHelloWorld contains whitelisted error: `[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.'
2020-11-11 23:51:20.811 INFO  [o.g.t.i.u.Logs] (checkLog) build-and-run.log log for micronautHelloWorld contains whitelisted error: `[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.'
2020-11-11 23:51:20.811 INFO  [o.g.t.i.u.Logs] (checkLog) build-and-run.log log for micronautHelloWorld contains whitelisted error: `[WARNING] Discovered module-info.class. Shading will break its strong encapsulation.'
2020-11-11 23:51:20.815 INFO  [o.g.t.i.u.Logs] (logMeasurements) 
App,buildTimeMs,timeToFirstOKRequestMs,RSSKb,FDs
MICRONAUT_HELLOWORLD,89100,130,64192,177
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 304.232 s - in org.graalvm.tests.integration.RuntimesSmokeTest
[INFO] Running org.graalvm.tests.integration.AppReproducersTest
2020-11-11 23:51:20.830 INFO  [o.g.t.i.AppReproducersTest] (nativeJVMTextProcessing) Testing app: DEBUG_SYMBOLS_SMOKE
2020-11-11 23:52:08.017 INFO  [o.g.t.i.AppReproducersTest] (nativeJVMTextProcessing) Running JVM mode...
2020-11-11 23:52:11.631 INFO  [o.g.t.i.AppReproducersTest] (nativeJVMTextProcessing) Running Native mode...
2020-11-11 23:52:18.942 INFO  [o.g.t.i.u.Logs] (logMeasurements) 
App,timeToFinishMs
DEBUG_SYMBOLS_SMOKE_JVM,3613
2020-11-11 23:52:18.942 INFO  [o.g.t.i.u.Logs] (logMeasurements) 
App,timeToFinishMs
DEBUG_SYMBOLS_SMOKE_NATIVE,7303
2020-11-11 23:52:18.986 INFO  [o.g.t.i.AppReproducersTest] (debugSymbolsQuarkus) Testing app: DEBUG_QUARKUS_FULL_MICROPROFILE
2020-11-11 23:55:05.728 INFO  [o.g.t.i.AppReproducersTest] (timezonesBakedIn) Testing app: TIMEZONES
2020-11-11 23:55:27.102 INFO  [o.g.t.i.AppReproducersTest] (timezonesBakedIn) Running...
2020-11-11 23:55:27.124 INFO  [o.g.t.i.AppReproducersTest] (debugSymbolsSmokeGDB) Testing app: DEBUG_SYMBOLS_SMOKE
2020-11-11 23:56:10.140 INFO  [o.g.t.i.AppReproducersTest] (randomNumbersReinit) Testing app: RANDOM_NUMBERS
2020-11-11 23:56:52.612 INFO  [o.g.t.i.AppReproducersTest] (randomNumbersReinit) Running...#1
2020-11-11 23:56:52.614 INFO  [o.g.t.i.AppReproducersTest] (randomNumbersReinit) Running...#2
2020-11-11 23:56:52.617 INFO  [o.g.t.i.AppReproducersTest] (randomNumbersReinit) [secureRandom: [4, 5, 53, 8, 64, 64, 45, 47, 95, 44], UUID: d6c99c34-0b64-40c6-9afa-4324a63a5c2b, secureRandom: [26, 96, 39, 75, 11, 32, 82, 4, 98, 43], UUID: afba2bee-884f-4685-8337-2958e678006d]
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 331.787 s - in org.graalvm.tests.integration.AppReproducersTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ testsuite ---
[INFO] Building jar: /home/karm/workspaceRH/mandrel-integration-tests/testsuite/target/testsuite-1.0.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-failsafe-plugin:2.22.2:integration-test (default) @ testsuite ---
[INFO] 
[INFO] --- maven-failsafe-plugin:2.22.2:verify (default) @ testsuite ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Native image integration TS 1.0.0-SNAPSHOT:
[INFO] 
[INFO] Native image integration TS ........................ SUCCESS [  0.099 s]
[INFO] testsuite .......................................... SUCCESS [10:38 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10:38 min
[INFO] Finished at: 2020-11-11T23:56:53+01:00
[INFO] ------------------------------------------------------------------------
```

# Testsuite builder image

```
arm@localhost:~/workspaceRH/mandrel-integration-tests (debug-symbols)$  mvn clean verify -Ptestsuite-builder-image -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel:20.2.0.0.Final-java11
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] Native image integration TS                                        [pom]
[INFO] testsuite                                                          [jar]
[INFO] 
[INFO] ----------------< org.graalvm.tests.integration:parent >----------------
[INFO] Building Native image integration TS 1.0.0-SNAPSHOT                [1/2]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ parent ---
[INFO] 
[INFO] --------------< org.graalvm.tests.integration:testsuite >---------------
[INFO] Building testsuite 1.0.0-SNAPSHOT                                  [2/2]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ testsuite ---
[INFO] Deleting /home/karm/workspaceRH/mandrel-integration-tests/testsuite/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ testsuite ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ testsuite ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- build-helper-maven-plugin:3.0.0:add-test-source (add-integration-test-source-as-test-sources) @ testsuite ---
[INFO] Test Source directory: /home/karm/workspaceRH/mandrel-integration-tests/testsuite/src/it/java added.
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ testsuite ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 0 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ testsuite ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 13 source files to /home/karm/workspaceRH/mandrel-integration-tests/testsuite/target/test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.22.2:test (default-test) @ testsuite ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.graalvm.tests.integration.RuntimesSmokeTest
2020-11-11 23:59:25.324 WARN  [o.g.t.i.u.Commands] (getProperty) Failed to detect any of QUARKUS_NATIVE_CONTAINER_RUNTIME,quarkus.native.container-runtime as env or sys props, defaulting to docker
2020-11-11 23:59:25.348 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Testing app: QUARKUS_BUILDER_IMAGE_ENCODING
2020-11-12 00:00:28.208 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Running...
2020-11-12 00:00:29.356 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Testing web page content...
2020-11-12 00:00:29.357 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Terminate and scan logs...
2020-11-12 00:00:31.629 INFO  [o.g.t.i.RuntimesSmokeTest] (testRuntime) Gonna wait for ports closed...
2020-11-12 00:00:31.652 INFO  [o.g.t.i.u.Logs] (logMeasurements) 
App,buildTimeMs,timeToFirstOKRequestMs,RSSKb
QUARKUS_BUILDER_IMAGE_ENCODING,62732,1142,35840
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 66.684 s - in org.graalvm.tests.integration.RuntimesSmokeTest
[INFO] Running org.graalvm.tests.integration.AppReproducersTest
2020-11-12 00:00:31.687 INFO  [o.g.t.i.AppReproducersTest] (debugSymbolsQuarkusContainer) Testing app: DEBUG_QUARKUS_BUILDER_IMAGE_VERTX
2020-11-12 00:02:30.919 INFO  [o.g.t.i.u.Logs] (checkLog) build-and-run.log log for debugSymbolsQuarkusContainer contains whitelisted error: `2020-11-11 23:02:28,590 WARN  [io.ver.sql.imp.SocketConnectionBase] (vert.x-eventloop-thread-22) Backend notice: severity='NOTICE', code='00000', message='table "fruits" does not exist, skipping', detail='null', hint='null', position='null', internalPosition='null', internalQuery='null', where='null', file='tablecmds.c', line='914', routine='DropErrorMsgNonExistent', schema='null', table='null', column='null', dataType='null', constraint='null''
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 119.396 s - in org.graalvm.tests.integration.AppReproducersTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ testsuite ---
[INFO] Building jar: /home/karm/workspaceRH/mandrel-integration-tests/testsuite/target/testsuite-1.0.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-failsafe-plugin:2.22.2:integration-test (default) @ testsuite ---
[INFO] 
[INFO] --- maven-failsafe-plugin:2.22.2:verify (default) @ testsuite ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Native image integration TS 1.0.0-SNAPSHOT:
[INFO] 
[INFO] Native image integration TS ........................ SUCCESS [  0.088 s]
[INFO] testsuite .......................................... SUCCESS [03:08 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:08 min
[INFO] Finished at: 2020-11-12T00:02:31+01:00
[INFO] ------------------------------------------------------------------------

```

**Comment by Karm** [comment]
The debug session automation is complete for Linux and GDB now. Merging and moving to Windows debug symbols automation....

**Comment by Karm** [comment]
@zakkak FYI...    I don't  want to demand your time to chew through this all... I should have proceeded more incrementally :)  

Perhaps the interesting parts:
 * [GDBSession.java](https://github.com/Karm/mandrel-integration-tests/blob/9066be5146f6321ae6b2a0e0ceb0d3483e47ece8/testsuite/src/it/java/org/graalvm/tests/integration/utils/GDBSession.java) "encodes" GDB sessions
 * [AppReproducersTest.java](https://github.com/Karm/mandrel-integration-tests/blob/9066be5146f6321ae6b2a0e0ceb0d3483e47ece8/testsuite/src/it/java/org/graalvm/tests/integration/AppReproducersTest.java) contains the actual GDB related tests (both locally and in container)
 * [BuildAndRunCmds.java](https://github.com/Karm/mandrel-integration-tests/blob/9066be5146f6321ae6b2a0e0ceb0d3483e47ece8/testsuite/src/it/java/org/graalvm/tests/integration/utils/BuildAndRunCmds.java#L53) starting apps and containers...

WTF it does:
 - GDB smoke test on a small app that parses text (the same app with huge input is also used to generate the time difference between Native svm and JVM Hotspot modes (Native is easy 2x slower). I would like to keep track of the gap, making sure it closes, not widens entually.
 - Quarkus full MicroProfle debugged with GDB locally, on the machine the test runs
 - Quarkus vert.x/Mutiny + Postgres db driver app debugged in container with gdb installed there (microdnf)

File: builder-image-apps/quarkus-vertx/src/main/docker/Dockerfile.native [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+#
+# docker run -i --rm -p 8080:8080 quarkus/using-vertx
+#
+###
+FROM registry.access.redhat.com/ubi8/ubi-minimal:8.2
+RUN microdnf --enablerepo=*debug* install gdb glibc-debuginfo zlib-debuginfo
+WORKDIR /work/
+COPY target/*-runner /work/application
+COPY target/*-runner.debug /work/
+COPY target/quarkus-native-image-source-jar/sources /work/sources
```
Comment (zakkak) [inline builder-image-apps/quarkus-vertx/src/main/docker/Dockerfile.native:11]: Why don't you keep the same folder hierarchy?

File: builder-image-apps/quarkus-vertx/src/main/resources/application.properties [hunk 2]
```diff
@@ -0,0 +1,8 @@
+quarkus.ssl.native=true
+
+quarkus.datasource.db-kind=postgresql
+quarkus.datasource.username=quarkus_test
+quarkus.datasource.password=quarkus_test
+quarkus.datasource.reactive.url=postgresql://localhost:5432/quarkus_test
```
Comment (zakkak) [inline builder-image-apps/quarkus-vertx/src/main/resources/application.properties:6]: Who's responsible for starting postgresql? Did I miss it?

**Review by zakkak — APPROVED** [review]
LGTM

File: builder-image-apps/quarkus-vertx/src/main/docker/Dockerfile.native (hunk 1 above)
Comment (Karm) [reply builder-image-apps/quarkus-vertx/src/main/docker/Dockerfile.native:11]: Hmm. Dunno. Gonna take a look...

**Pull request merged**